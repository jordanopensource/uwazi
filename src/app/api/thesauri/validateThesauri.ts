import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';
import { wrapValidator } from 'shared/tsUtils';
import { thesaurusSchema } from 'shared/types/thesaurusSchema';
import { ThesaurusSchema } from 'shared/types/thesaurusType';
import model from './dictionariesModel';

const ajv = ajvKeywords(new Ajv({ allErrors: true }), ['uniqueItemProperties']);
ajv.addVocabulary(['tsType']);

ajv.addKeyword({
  keyword: 'uniqueName',
  async: true,
  validate: async (_config: any, thesaurus: ThesaurusSchema) => {
    const [duplicated] = await model.get({
      _id: { $ne: thesaurus._id },
      name: new RegExp(`^${thesaurus.name}$` || '', 'i'),
    });

    if (duplicated) {
      return false;
    }
    return true;
  },
});

const getDuplicatedLabels = (values: { label: string }[] | undefined): string[] => {
  if (!values) return [];
  const labels = values.map(v => v.label);
  const asSet: Set<string> = new Set();
  const duplicated: Set<string> = new Set();
  labels.forEach(label => {
    if (asSet.has(label)) {
      duplicated.add(label);
    }
    asSet.add(label);
  });
  return Array.from(duplicated);
};

ajv.addKeyword({
  keyword: 'uniqueLabels',
  validate: (_config: any, thesaurus: ThesaurusSchema) => {
    const duplicated = getDuplicatedLabels(thesaurus.values);
    thesaurus.values?.forEach(v => {
      duplicated.push(...getDuplicatedLabels(v.values).map(l => `${v.label}/${l}`));
    });
    if (duplicated.length > 0) {
      throw new Ajv.ValidationError([
        {
          instancePath: '',
          schemaPath: '#/uniqueLabels',
          keyword: 'uniqueLabels',
          params: {
            duplicated,
          },
          message: `Duplicated labels: ${duplicated.join(', ')}.`,
        },
      ]);
    }
    return true;
  },
});

export const validateThesauri = wrapValidator(ajv.compile(thesaurusSchema));
