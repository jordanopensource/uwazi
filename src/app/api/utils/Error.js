/**
 * @deprecated
 */
export default function (message, code = 500, logLevel = 'debug') {
  if (code === 500) {
    logLevel = 'error';
  }

  if (message.ajv) {
    return { message: message.message, ...message, code, logLevel };
  }

  if (message instanceof Error) {
    return { message: message.message, stack: message.stack, code, original: message, logLevel };
  }

  return { message, code, logLevel };
}
