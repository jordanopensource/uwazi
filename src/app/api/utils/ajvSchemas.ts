const ObjectIdAsString = {
  type: 'string',
  pattern: '^[0-9a-fA-F]{24}$',
};

export { ObjectIdAsString };
