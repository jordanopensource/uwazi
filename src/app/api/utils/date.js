import moment from 'moment';
// @ts-ignore
import parser from 'any-date-parser';
import * as stopword from 'stopword';

export default {
  currentUTC() {
    return moment.utc().toDate().getTime();
  },

  descriptionToTimestamp(date) {
    if (date === 'last-day-last-month') {
      return moment.utc().subtract(1, 'months').endOf('month').unix();
    }

    if (date === 'first-day-last-month') {
      return moment.utc().subtract(1, 'months').startOf('month').unix();
    }

    return date;
  },

  addYearsToCurrentDate(yearsToAdd) {
    const newDate = new Date();
    newDate.setHours(0, 0, 0, 0);
    newDate.setFullYear(newDate.getFullYear() + yearsToAdd);
    return newDate;
  },

  dateToSeconds(value, locale) {
    let parsedValue = value.normalize('NFD').replace(/[\u0300-\u036f]/g, '');

    const { removeStopwords, ...languages } = stopword;

    if (locale) {
      parsedValue = removeStopwords(parsedValue.split(' '), languages[locale]).join(' ');
    }

    let getDate = parser.fromString(parsedValue, locale);
    if (getDate.invalid) {
      getDate = Date.parse(`${parsedValue} GMT`);
    }
    const formattedDate = getDate / 1000;
    return formattedDate;
  },
};
