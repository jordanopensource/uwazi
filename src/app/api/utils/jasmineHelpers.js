/** @format */

const assessStatus = routeResult => {
  let status;
  const req = {};
  const next = () => {};
  const res = {
    status: code => {
      status = code;
    },
    json: () => {},
  };

  routeResult.middlewares[0](req, res, next);

  return status;
};

const assessAuthorized = (routeResult, roles = []) => {
  const nextCalledFor = [];

  roles.forEach(role => {
    const req = {
      user: { role },
      get: key => (key === 'X-Requested-With' ? 'XMLHttpRequest' : ''),
    };
    const res = { status: () => {}, json: () => {} };
    const next = () => {
      nextCalledFor.push(role);
    };
    routeResult.middlewares[0](req, res, next);
  });

  return roles.join(',') === nextCalledFor.join(',');
};

const conformValidationErrors = (routeCanFail, expected) =>
  routeCanFail
    ? `Route is not correctly authorized for ['${expected.join("', '")}']`
    : 'Route is not authorized ! (Auth middleware should be the first one)';

const conformMessage = (result, routeCanFail, expected) =>
  result.pass ? 'route is authorized' : conformValidationErrors(routeCanFail, expected);

const matchers = {
  toNeedAuthorization(routeResult, expected) {
    const routeCanFail = assessStatus(routeResult) === 401;
    const routeValidatesExpected = expected ? assessAuthorized(routeResult, expected) : true;
    const result = { pass: routeCanFail && routeValidatesExpected };
    result.message = () => conformMessage(result, routeCanFail, expected);
    return result;
  },
};

expect.extend(matchers);

export function catchErrors(done) {
  return error => {
    if (error instanceof Error) {
      return done.fail(error);
    }
    return done.fail(JSON.stringify(error));
  };
}
