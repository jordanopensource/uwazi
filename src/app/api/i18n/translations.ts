import { CSVLoader } from 'api/csv';
import { generateFileName } from 'api/files';
import { CreateTranslationsData } from 'api/i18n.v2/services/CreateTranslationsService';
import { DefaultTranslations } from 'api/i18n/defaultTranslations';
import { legacyLogger } from 'api/log';
import { EnforcedWithId, WithId } from 'api/odm';
import settings from 'api/settings/settings';
import thesauri from 'api/thesauri/thesauri';
import { prettifyError } from 'api/utils/handleError';
import * as os from 'os';
import path from 'path';
import { TranslationContext, TranslationType, TranslationValue } from 'shared/translationType';
// eslint-disable-next-line node/no-restricted-import
import { createWriteStream } from 'fs';
import { ObjectId } from 'mongodb';
import { availableLanguages } from 'shared/languagesList';
import { ContextType } from 'shared/translationSchema';
import { LanguageISO6391 } from 'shared/types/commonTypes';
import { pipeline } from 'stream/promises';
import {
  addLanguageV2,
  deleteTranslationsByContextIdV2,
  deleteTranslationsByLanguageV2,
  getTranslationsV2,
  getTranslationsV2ByContext,
  getTranslationsV2ByLanguage,
  updateContextV2,
  upsertTranslationsV2,
} from './v2_support';

function checkForMissingKeys(
  keyValuePairsPerLanguage: { [x: string]: { [k: string]: string } },
  translation: WithId<TranslationType>,
  valueDict: IndexedContextValues,
  contextId: string
) {
  if (!translation.locale) throw new Error('Translation local does not exist !');
  const missingKeys = Object.keys(keyValuePairsPerLanguage[translation.locale]).filter(
    key => !(key in valueDict)
  );
  if (missingKeys.length) {
    throw new Error(
      `Process is trying to update missing translation keys: ${translation.locale} - ${contextId} - ${missingKeys}.`
    );
  }
}

function prepareContexts(contexts: TranslationContext[] = []) {
  return contexts.map(context => ({
    ...context,
    type:
      context.id === 'System' || context.id === 'Filters' || context.id === 'Menu'
        ? 'Uwazi UI'
        : context.type,
    values: context.values
      ? context.values.reduce((values, value) => {
          if (value.key && value.value) {
            values[value.key] = value.value; //eslint-disable-line no-param-reassign
          }
          return values;
        }, {} as IndexedContextValues)
      : {},
  }));
}

function checkDuplicateKeys(
  context: TranslationContext | IndexedContext,
  values: TranslationValue[]
) {
  if (!values) return;

  const seen = new Set();
  values.forEach(value => {
    if (seen.has(value.key)) {
      throw new Error(
        `Process is trying to save repeated translation key ${value.key} in context ${context.id} (${context.type}).`
      );
    }
    seen.add(value.key);
  });
}

function processContextValues(context: TranslationContext | IndexedContext): TranslationContext {
  const processedValues: TranslationValue[] = [];

  if (context.values && !Array.isArray(context.values)) {
    const indexedValues: IndexedContextValues = context.values;
    Object.keys(indexedValues).forEach(key => {
      if (indexedValues[key]) {
        processedValues.push({ key, value: indexedValues[key] });
      }
    });
  }

  let values: TranslationValue[] = [];
  if (processedValues.length) {
    values = processedValues;
  }
  if (Array.isArray(context.values)) {
    values = context.values as TranslationValue[];
  }

  checkDuplicateKeys(context, values);

  return { ...context, values };
}

const propagateTranslation = async (
  translation: TranslationType,
  currentTranslationData: WithId<TranslationType>
) => {
  await (currentTranslationData.contexts || ([] as TranslationContext[])).reduce(
    async (promise: Promise<any>, context) => {
      await promise;

      const isPresentInTheComingData = (translation.contexts || []).find(
        _context => _context.id?.toString() === context.id?.toString()
      );

      if (isPresentInTheComingData && isPresentInTheComingData.type === 'Thesaurus') {
        const thesaurus = await thesauri.getById(context.id);

        const valuesChanged: IndexedContextValues = (isPresentInTheComingData.values || []).reduce(
          (changes, value) => {
            const currentValue = (context.values || []).find(v => v.key === value.key);
            if (currentValue?.key && currentValue.value !== value.value) {
              return { ...changes, [currentValue.key]: value.value } as IndexedContextValues;
            }
            return changes;
          },
          {} as IndexedContextValues
        );

        const changesMatchingDictionaryId = Object.keys(valuesChanged)
          .map(valueChanged => {
            const valueFound = (thesaurus?.values || []).find(v => v.label === valueChanged);
            if (valueFound?.id) {
              return { id: valueFound.id, value: valuesChanged[valueChanged] };
            }
            return null;
          })
          .filter(a => a) as { id: string; value: string }[];

        return Promise.all(
          changesMatchingDictionaryId.map(async change =>
            thesauri.renameThesaurusInMetadata(
              change.id,
              change.value,
              context.id,
              translation.locale
            )
          )
        );
      }
      return Promise.resolve([]);
    },
    Promise.resolve([])
  );
};

const translationTypeToIndexedTranslation = (translations?: EnforcedWithId<TranslationType>[]) =>
  translations
    ? translations.map(
        translation =>
          ({
            ...translation,
            contexts: prepareContexts(translation.contexts),
          }) as IndexedTranslations
      )
    : [];

export default {
  prepareContexts,

  async get(query: { locale?: LanguageISO6391; context?: string } = {}) {
    const language = query.locale;

    if (query.context) {
      return translationTypeToIndexedTranslation(await getTranslationsV2ByContext(query.context));
    }

    if (language) {
      return translationTypeToIndexedTranslation(await getTranslationsV2ByLanguage(language));
    }

    return translationTypeToIndexedTranslation(await getTranslationsV2());
  },

  async save(translation: TranslationType | IndexedTranslations) {
    const translationToSave = {
      ...translation,
      contexts: translation.contexts && translation.contexts.map(processContextValues),
    } as TranslationType;

    if (!translation.locale) {
      throw new Error('translation to save should have a locale');
    }

    const [currentTranslationData] = await getTranslationsV2ByLanguage(translation.locale);

    const processedTranslation: TranslationType & { contexts: TranslationContext[] } = {
      ...translation,
      contexts: (translation.contexts || []).map(processContextValues),
    };

    await propagateTranslation(processedTranslation, currentTranslationData);

    await upsertTranslationsV2([translationToSave]);
    return translationToSave;
  },

  async updateEntries(
    contextId: string,
    keyValuePairsPerLanguage: {
      [x: string]: { [k: string]: string };
    }
  ) {
    const { languages = [] } = await settings.get({}, 'languages');
    const languagesSet = new Set(languages.map(l => String(l.key)));

    const languagesToUpdate = Object.keys(keyValuePairsPerLanguage).filter(l =>
      languagesSet.has(l)
    ) as LanguageISO6391[];

    const translationsToUpdate = await Promise.all(
      languagesToUpdate.map(async language => {
        const [translation] = await getTranslationsV2ByLanguage(language);
        return translation;
      })
    );

    return Promise.all(
      translationsToUpdate.map(async translation => {
        if (!translation.locale) throw new Error('Translation local does not exist !');

        const context = (translation.contexts || []).find(c => c.id === contextId);
        if (!context) {
          return Promise.resolve();
        }
        const valueDict: IndexedContextValues = Object.fromEntries(
          (context.values || []).map(({ key, value }) => [key, value])
        );
        checkForMissingKeys(keyValuePairsPerLanguage, translation, valueDict, contextId);
        Object.entries(keyValuePairsPerLanguage[translation.locale]).forEach(([key, value]) => {
          valueDict[key] = value;
        });
        context.values = Object.entries(valueDict).map(([key, value]) => ({ key, value }));
        return this.save(translation);
      })
    );
  },

  async addContext(
    id: string | ObjectId,
    contextName: string,
    values: IndexedContextValues,
    type: ContextType
  ) {
    const translatedValues: TranslationValue[] = [];
    Object.keys(values).forEach(key => {
      translatedValues.push({ key, value: values[key] });
    });

    const result = await getTranslationsV2();
    await upsertTranslationsV2(
      result.map(translation => {
        // eslint-disable-next-line no-param-reassign
        translation.contexts = translation.contexts || [];
        translation.contexts.push({
          id: id.toString(),
          label: contextName,
          values: translatedValues,
          type,
        });
        return translation;
      })
    );

    return 'ok';
  },

  async deleteContext(contextId: string) {
    await deleteTranslationsByContextIdV2(contextId);
    return 'ok';
  },

  async updateContext(
    context: CreateTranslationsData['context'],
    keyNamesChanges: { [x: string]: string },
    deletedProperties: string[],
    values: IndexedContextValues
  ) {
    await updateContextV2(context, keyNamesChanges, deletedProperties, values);
    return 'ok';
  },

  async addLanguage(newLanguage: LanguageISO6391) {
    const [translations] = await getTranslationsV2ByLanguage(newLanguage);
    if (translations.contexts?.length) {
      return Promise.resolve();
    }

    const defaultLanguage = await settings.getDefaultLanguage();

    return addLanguageV2(newLanguage, defaultLanguage.key);
  },

  async removeLanguage(locale: LanguageISO6391) {
    return deleteTranslationsByLanguageV2(locale);
  },

  async importPredefined(locale: string) {
    const translationsCsv = await DefaultTranslations.retrievePredefinedTranslations(locale);
    const tmpCsv = path.join(os.tmpdir(), generateFileName({ originalname: 'tmp-csv.csv' }));
    await pipeline(translationsCsv, createWriteStream(tmpCsv));
    const loader = new CSVLoader();
    await loader.loadTranslations(tmpCsv, 'System');
  },

  async availableLanguages() {
    let languagesWithTranslations: string[] = [];
    try {
      languagesWithTranslations = await DefaultTranslations.retrieveAvailablePredefinedLanguages();
    } catch (e) {
      legacyLogger.error(prettifyError(e));
      return availableLanguages;
    }
    return availableLanguages.map(language => ({
      ...language,
      translationAvailable: languagesWithTranslations.includes(language.key),
    }));
  },
};

export interface IndexedContextValues {
  [k: string]: string;
}

export interface IndexedContext extends Omit<TranslationContext, 'values'> {
  values: IndexedContextValues;
}

export interface IndexedTranslations extends Omit<TranslationType, 'contexts'> {
  contexts?: IndexedContext[];
}
