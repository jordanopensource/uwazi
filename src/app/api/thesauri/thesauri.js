import _ from 'lodash';
import {
  generateIds,
  getUpdatedIds,
  getUpdatedNames,
  getDeletedProperties,
} from 'api/templates/utils';
import entities from 'api/entities/entities';
import { preloadOptionsLimit } from 'shared/config';
import templates from 'api/templates/templates';
import settings from 'api/settings/settings';
import translations from 'api/i18n/translations';
import { denormalizeThesauriLabelInMetadata } from 'api/entities/denormalize';
import { search } from 'api/search';
import model from './dictionariesModel';
import { validateThesauri } from './validateThesauri';
import { objectIndex } from 'shared/data_utils/objectIndex';

const autoincrementValuesId = thesauri => {
  thesauri.values = generateIds(thesauri.values);

  thesauri.values = thesauri.values.map(value => {
    if (value.values) {
      value.values = generateIds(value.values);
    }

    return value;
  });
  return thesauri;
};

function normalizeThesaurusLabel(label) {
  const trimmed = label.trim().toLowerCase();
  return trimmed.length > 0 ? trimmed : null;
}

function thesauriToTranslationContext(thesauri) {
  return thesauri.values.reduce((ctx, prop) => {
    ctx[prop.label] = prop.label;
    if (prop.values) {
      const propctx = prop.values.reduce((_ctx, val) => {
        _ctx[val.label] = val.label;
        return _ctx;
      }, {});
      ctx = Object.assign(ctx, propctx);
    }
    return ctx;
  }, {});
}

const create = async thesauri => {
  const context = thesauriToTranslationContext(thesauri);
  context[thesauri.name] = thesauri.name;

  const created = await model.save(thesauri);
  await translations.addContext(created._id, thesauri.name, context, 'Thesaurus');
  return created;
};

const updateTranslation = (current, thesauri) => {
  const currentProperties = current.values;
  const newProperties = thesauri.values;

  const { update: updatedLabels, delete: removedThroughUpdate } = getUpdatedNames(
    {
      prop: 'label',
      outKey: 'label',
      filterBy: 'id',
    },
    currentProperties,
    newProperties
  );
  if (current.name !== thesauri.name) {
    updatedLabels[current.name] = thesauri.name;
  }
  const deletedPropertiesByLabel = getDeletedProperties(
    currentProperties,
    newProperties,
    'id',
    'label'
  );
  const allRemoved = Array.from(new Set(deletedPropertiesByLabel.concat(removedThroughUpdate)));

  const context = thesauriToTranslationContext(thesauri);

  context[thesauri.name] = thesauri.name;
  return translations.updateContext(
    { id: current._id.toString(), label: thesauri.name, type: 'Thesaurus' },
    updatedLabels,
    allRemoved,
    context
  );
};

async function updateOptionsInEntities(current, thesauri) {
  const currentProperties = current.values;
  const newProperties = thesauri.values;
  const deletedPropertiesById = getDeletedProperties(currentProperties, newProperties, 'id', 'id');
  await Promise.all(
    deletedPropertiesById.map(deletedId =>
      entities.deleteThesaurusFromMetadata(deletedId, thesauri._id)
    )
  );

  const updatedIds = getUpdatedIds(
    {
      prop: 'label',
      filterBy: 'id',
    },
    currentProperties,
    newProperties
  );
  const toUpdate = [];

  Object.keys(updatedIds).forEach(id => {
    const option = newProperties
      .reduce((flattendedOptions, o) => flattendedOptions.concat([o, ...(o.values || [])]), [])
      .find(o => o.id === id);

    if (option.values?.length) {
      option.values.forEach(o => {
        toUpdate.push({ id: o.id, label: o.label, parent: { id, label: updatedIds[id] } });
      });
      return;
    }

    toUpdate.push({ id, label: updatedIds[id] });
  });

  const defaultLanguage = (await settings.get()).languages.find(lang => lang.default).key;
  await Promise.all(
    toUpdate.map(option =>
      denormalizeThesauriLabelInMetadata(
        option.id,
        option.label,
        thesauri._id,
        defaultLanguage,
        option.parent
      )
    )
  );
}

const update = async thesauri => {
  const currentThesauri = await model.getById(thesauri._id);
  const valuesHaveChanged =
    JSON.stringify(thesauri.values) !== JSON.stringify(currentThesauri.values);
  const nameHasChanged = thesauri.name !== currentThesauri.name;
  if (valuesHaveChanged || nameHasChanged) {
    await updateTranslation(currentThesauri, thesauri);
    await updateOptionsInEntities(currentThesauri, thesauri);
  }
  return model.save(thesauri);
};

function calcNewLabels(originals, news) {
  const originalLabels = originals.map(v => v.label);
  const normalizedOriginals = originalLabels.map(normalizeThesaurusLabel);
  const normalizedSet = new Set(normalizedOriginals);
  const actualNewLabels = [];
  news.forEach(({ label }) => {
    const normalized = normalizeThesaurusLabel(label);
    if (!normalizedSet.has(normalized)) {
      actualNewLabels.push(label);
      normalizedSet.add(normalized);
    }
  });
  return actualNewLabels.map(label => ({ label }));
}

function calcNewValues(originalValues, newValues) {
  const values = _.cloneDeep(originalValues);
  const roots = values.filter(v => !v.values);
  const groups = values.filter(v => v.values);
  const [newRoots, newGroups] = _.partition(newValues, v => !v.values);

  const finalNewRoots = calcNewLabels(roots, newRoots);
  values.push(...finalNewRoots);

  const groupsByNormalizedLabel = objectIndex(
    groups,
    v => normalizeThesaurusLabel(v.label),
    v => v
  );
  const finalNewGroups = [];
  newGroups.forEach(newGroup => {
    const normalizedLabel = normalizeThesaurusLabel(newGroup.label);
    if (!(normalizedLabel in groupsByNormalizedLabel)) {
      const emptyNewGroup = { label: newGroup.label, values: [] };
      finalNewGroups.push(emptyNewGroup);
      groupsByNormalizedLabel[normalizedLabel] = emptyNewGroup;
    }
    const group = groupsByNormalizedLabel[normalizedLabel];
    const newLocalValues = calcNewLabels(group.values, newGroup.values);
    group.values.push(...newLocalValues);
  });
  values.push(...finalNewGroups);

  return values;
}

const thesauri = {
  async save(t) {
    const toSave = { values: [], type: 'thesauri', ...t };

    autoincrementValuesId(toSave);

    await validateThesauri(toSave);

    if (toSave._id) {
      return update(toSave);
    }
    return create(toSave);
  },

  appendValues(thesaurus, newValues) {
    return {
      ...thesaurus,
      values: calcNewValues(thesaurus.values || [], newValues),
    };
  },

  entitiesToThesauri(_entities) {
    const values = _entities.map(entity => ({
      id: entity.sharedId,
      label: entity.title,
      icon: entity.icon,
    }));
    return { values };
  },

  async templateToThesauri(template, language, user, countPerTemplate) {
    const onlyPublished = !user;
    const _entities = await entities.getByTemplate(
      template._id,
      language,
      preloadOptionsLimit(),
      onlyPublished
    );
    const values = this.entitiesToThesauri(_entities);
    return Object.assign(template, values, {
      type: 'template',
      optionsCount: countPerTemplate[template._id.toString()],
    });
  },

  getById(id) {
    return model.getById(id);
  },

  async get(thesauriId, language, user) {
    let query;
    if (thesauriId) {
      query = { _id: thesauriId };
    }

    const dictionaries = await model.get(query);
    const allTemplates = await templates.get(query);

    if (allTemplates.length && language) {
      const templateCount = await search.countPerTemplate(language);

      const processedTemplates = await Promise.all(
        allTemplates.map(result =>
          this.templateToThesauri(result, language, user, templateCount).then(
            templateTransformedInThesauri => templateTransformedInThesauri
          )
        )
      );
      return dictionaries.concat(processedTemplates);
    }

    return dictionaries;
  },

  dictionaries(query) {
    return model.get(query);
  },

  delete(id) {
    return templates
      .countByThesauri(id)
      .then(count => {
        if (count) {
          return Promise.reject({ key: 'templates_using_dictionary', value: count });
        }
        return translations.deleteContext(id);
      })
      .then(() => model.delete(id))
      .then(() => ({ ok: true }));
  },

  async renameThesaurusInMetadata(valueId, newLabel, thesaurusId, language) {
    return denormalizeThesauriLabelInMetadata(valueId, newLabel, thesaurusId, language);
  },
};

const flatThesaurusValues = (thesaurus, includeRoots = false) =>
  includeRoots
    ? _.flatMapDeep(thesaurus?.values, tv => {
        const { values = [], ...root } = tv;
        const valuesCopy = Array.from(values);
        valuesCopy.push(root);
        return valuesCopy;
      })
    : _.flatMapDeep(thesaurus?.values, tv => tv.values || tv);

export default thesauri;
export { thesauri, flatThesaurusValues, normalizeThesaurusLabel };
