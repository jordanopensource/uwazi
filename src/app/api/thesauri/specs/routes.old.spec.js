import 'api/utils/jasmineHelpers';
import db from 'api/utils/testing_db';
import translations from 'api/i18n/translations';

import { fixtures } from './fixtures';
import instrumentRoutes from '../../utils/instrumentRoutes';
import thesauri from '../thesauri';
import thesauriRoute from '../routes.js';

describe('thesauri routes', () => {
  let routes;

  beforeEach(async () => {
    routes = instrumentRoutes(thesauriRoute);
    await db.setupFixturesAndContext(fixtures);
  });

  afterAll(async () => {
    await db.disconnect();
  });

  describe('GET', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/thesauris')).toMatchSnapshot();
    });

    it('should return all thesauri by default, passing user', async () => {
      jest.spyOn(thesauri, 'get').mockImplementation(async () => Promise.resolve('response'));
      const response = await routes.get('/api/thesauris', { language: 'es', user: 'user' });
      let undefinedVar;
      expect(thesauri.get).toHaveBeenCalledWith(undefinedVar, 'es', 'user');
      expect(response).toEqual({ rows: 'response' });
    });

    describe('when passing id', () => {
      it('should get passing id', async () => {
        jest.spyOn(thesauri, 'get').mockImplementation(async () => Promise.resolve('response'));
        const req = { query: { _id: 'id' } };

        await routes.get('/api/thesauris', req);
        let undefinedVar;
        expect(thesauri.get).toHaveBeenCalledWith('id', undefinedVar, undefinedVar);
      });
    });

    describe('dictionaries', () => {
      it('should have a validation schema', () => {
        expect(routes.get.validation('/api/dictionaries')).toMatchSnapshot();
      });

      it('should return all dictionaries by default', async () => {
        jest
          .spyOn(thesauri, 'dictionaries')
          .mockImplementation(async () => Promise.resolve('response'));
        const response = await routes.get('/api/dictionaries');
        expect(thesauri.dictionaries).toHaveBeenCalled();
        expect(response).toEqual({ rows: 'response' });
      });

      describe('when passing id', () => {
        it('should get matching id', async () => {
          jest
            .spyOn(thesauri, 'dictionaries')
            .mockImplementation(async () => Promise.resolve('response'));
          const response = await routes.get('/api/dictionaries', { query: { _id: 'id' } });
          expect(thesauri.dictionaries).toHaveBeenCalledWith({ _id: 'id' });
          expect(response).toEqual({ rows: 'response' });
        });
      });
    });
  });

  describe('DELETE', () => {
    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/thesauris')).toMatchSnapshot();
    });

    it('should delete a thesauri', async () => {
      jest.spyOn(thesauri, 'delete').mockImplementation(async () => Promise.resolve());
      const req = { query: { _id: 'abc', _rev: '123' } };
      await routes.delete('/api/thesauris', req);
      expect(thesauri.delete).toHaveBeenCalledWith('abc', '123');
    });
  });

  describe('POST', () => {
    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/thesauris')).toMatchSnapshot();
    });

    it('should create a thesauri', async () => {
      jest.spyOn(translations, 'addContext').mockImplementation(async () => Promise.resolve());
      const req = { body: { name: 'Batman wish list', values: [{ id: '1', label: 'Joker BFF' }] } };
      const response = await routes.post('/api/thesauris', req);
      expect(response.values[0].id).toEqual('1');
      expect(response.values[0].label).toEqual('Joker BFF');
    });
  });
});
