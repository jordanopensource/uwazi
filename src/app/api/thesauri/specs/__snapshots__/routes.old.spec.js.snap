// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thesauri routes DELETE should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "_id": Object {
          "type": "string",
        },
        "_rev": Object {
          "type": "string",
        },
      },
      "required": Array [
        "_id",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`thesauri routes GET dictionaries should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "_id": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`thesauri routes GET should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "_id": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`thesauri routes POST should have a validation schema 1`] = `
Object {
  "properties": Object {
    "body": Object {
      "anyOf": Array [
        Object {
          "properties": Object {
            "__v": Object {
              "type": "number",
            },
            "_id": Object {
              "type": "string",
            },
            "enable_classification": Object {
              "type": "boolean",
            },
            "name": Object {
              "type": "string",
            },
            "values": Object {
              "items": Object {
                "properties": Object {
                  "_id": Object {
                    "type": "string",
                  },
                  "id": Object {
                    "type": "string",
                  },
                  "label": Object {
                    "type": "string",
                  },
                  "values": Object {
                    "items": Object {
                      "type": "object",
                    },
                    "type": "array",
                  },
                },
                "required": Array [
                  "label",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": Array [
            "name",
            "values",
          ],
          "type": "object",
        },
        Object {
          "properties": Object {
            "thesauri": Object {
              "type": "string",
            },
          },
          "required": Array [
            "thesauri",
          ],
          "type": "object",
        },
      ],
    },
  },
  "required": Array [
    "body",
  ],
  "type": "object",
}
`;
